File: CpuX64/disassembler_tool.py
#!/usr/bin/python3

import sys

from CpuX64 import symbolic
from CpuX64 import opcode_tab as x64


def disass(data):
    ins = x64.Disassemble(data)
    if ins is None:
        print(f"could not determine opcode {x64.Hexify(data)}")
        return

    enum_name, ops_str = symbolic.InsSymbolizeObjdumpCompat(ins, False)
    print(f"{x64.Hexify(data)}", f"{ins.opcode.name}_{ins.opcode.variant} {' '.join(ops_str)}")

    enum_name, ops_str = symbolic.InsSymbolize(ins)
    print("    " + enum_name)
    for f, o, o_str in zip(ins.opcode.fields, ins.operands, ops_str):
        if isinstance(f, x64.OK):
            f = f.name
        print(f"    {f:35s} {o_str:10} (0x{o:x})")
    print()
    data2 = x64.Assemble(ins)
    assert data == data2, f"{data} vs {data2}"
    ins2 = symbolic.InsFromSymbolized(enum_name, ops_str)
    assert tuple(ins.operands) == tuple(
        ins2.operands), f"{ins.operands} vs {ins2.operands}"


def HexToData(s: str):
    return [int(x, 16) for x in s.split()]


def batch():
    for line in sys.stdin:
        line = line.split("#")[0].strip()
        if not line.strip(): continue
        data = HexToData(line.strip())
        ins = x64.Disassemble(data)
        if ins.opcode is None:
            print(f"could not determine opcode [{x64.Hexify(data)}]")
            continue
        enum_name, ops_str = symbolic.InsSymbolize(ins)
        print(f"{x64.Hexify(data)}", f"{ins.opcode.name}_{ins.opcode.variant} {' '.join(ops_str)}")

        data2 = x64.Assemble(ins)
        assert data == data2
        ins2 = symbolic.InsFromSymbolized(enum_name, ops_str)
        assert tuple(ins.operands) == tuple(
            ins2.operands), f"{ins.operands} vs {ins2.operands}"


if __name__ == "__main__":
    if len(sys.argv) > 1:
        if sys.argv[1] == "batch":
            batch()
        else:
            for seq in sys.argv[1:]:
                disass(HexToData(seq))

File: CpuX64/disassembler_tool.cc
// (c) Robert Muth - see LICENSE for more info

#include "CpuX64/opcode_gen.h"
#include "CpuX64/symbolic.h"
#include "Util/assert.h"
#include "Util/parse.h"

#include <cstdlib>
#include <iomanip>
#include <iostream>
#include <string_view>

using namespace cwerg::x64;

std::string ExtractData(std::string_view line) {
  std::string out;
  out.reserve(line.size());
  unsigned nibble;
  bool have_nibble = false;
  for (char c : line) {
    if (cwerg::IsWhiteSpace(c)) {
      if (have_nibble) {
        out.push_back(nibble);
        have_nibble = false;
      }
    } else {
      const int hex_digit = cwerg::HexDigit(c);
      ASSERT(hex_digit >= 0, "");
      if (have_nibble) {
        out.push_back((nibble << 4) | unsigned(hex_digit));
        have_nibble = false;
      } else {
        nibble = hex_digit;
        have_nibble = true;
      }
    }
  }
  if (have_nibble) {
    out.push_back(nibble);
    have_nibble = false;
  }
  return out;
}

void batch(std::string_view data, const std::string& line) {
  Ins ins;
  if (!Disassemble(&ins, data)) {
    std::cout << "could not find opcode for: " << line << "\n";
    return;
  }
  std::vector<std::string> ops;
  std::string_view enum_name = InsSymbolize(ins, false, false, &ops);
  std::cout << std::setw(30) << data << std::dec << " " << enum_name;
  std::string_view sep = " ";
  for (const std::string& op : ops) {
    std::cout << sep << op;
    sep = ", ";
  }
  std::cout << "\n";
}

void disass(std::string_view data, const std::string& line) {
  Ins ins;
  if (!Disassemble(&ins, data)) {
    std::cout << "could not disassemble " << line << std::dec << "\n";
    return;
  }
  std::vector<std::string> ops;
  std::string_view enum_name = InsSymbolize(ins, true, true, &ops);
  std::cout << line << " " << enum_name;
  std::string_view sep = " ";
  for (const std::string& op : ops) {
    std::cout << sep << op;
    sep = " ";
  }
  std::cout << "\n";

  ops.clear();
  enum_name = InsSymbolize(ins, false, false, &ops);
  std::cout << "    " << enum_name << "\n";
  for (unsigned x = 0; x < ins.opcode->num_fields; ++x) {
    std::cout << "    " << std::left << std::setw(35)
              << EnumToString(ins.opcode->fields[x]) << " " << std::setw(10)
              << ops[x] << " (" << "0x" << std::hex << int64_t (ins.operands[x]) << std::dec << ")\n";
  }
  std::cout << "\n";
#if 0
  // check that the assembler works - this is not strictly
      // necessary but useful for debugging the assembler
      const uint32_t data2 = Assemble(ins);
      if (data != data2) {
        std::cout << "Disassembler failure " << std::hex << data << " vs "
                  << data2 << "\n";
        return 1;
      }
      ASSERT(data == data2, "");
#endif
}

int main(int argc, char* argv[]) {
  if (argc <= 1) {
    std::cout << "no command specified\n";
    return 1;
  }
  if (std::string_view("batch") == argv[1]) {
    for (std::string line; getline(std::cin, line);) {
      if (line[0] == '#') continue;
      const std::string data = ExtractData(line);
      batch(data, line);
    }
  } else {
    std::string out;
    out.reserve(argc);
    for (unsigned i = 1; i < argc; ++i) {
      const std::string data = ExtractData(argv[i]);
      disass(data, argv[i]);
    }

    return 0;
  }
}
